### PR TITLE
SONAR-9148 fix partition size for large inputs in MySQL

### PR DESCRIPTION
**Pull request by bartfastiel**

### CODE REVIEW

File: server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java [hunk 1]
```diff
@@ -142,7 +142,7 @@ public int countUsersWithGlobalPermissionExcludingUserPermission(DbSession dbSes
   public Collection<Integer> keepAuthorizedUsersForRoleAndProject(DbSession dbSession, Collection<Integer> userIds, String role, long projectId) {
     return executeLargeInputs(
       userIds,
-      partitionOfIds -> mapper(dbSession).keepAuthorizedUsersForRoleAndProject(role, projectId, partitionOfIds));
+      partitionOfIds -> mapper(dbSession).keepAuthorizedUsersForRoleAndProject(role, projectId, partitionOfIds));//FIXME chunk size
```
Comment (SonarTech) [inline server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)
MINOR Move this trailing comment on the previous empty line. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)

File: server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java [hunk 2]
```diff
@@ -131,7 +131,7 @@ public int countUsersWithGlobalPermissionExcludingUserPermission(DbSession dbSes
         if (userId == null) {
           return mapper(dbSession).keepAuthorizedProjectIdsForAnonymous(role, componentIds);
         }
-        return mapper(dbSession).keepAuthorizedProjectIdsForUser(userId, role, componentIds);
+        return mapper(dbSession).keepAuthorizedProjectIdsForUser(userId, role, componentIds);//FIXME chunk size
```
Comment (SonarTech) [inline server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)
MINOR Move this trailing comment on the previous empty line. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java [hunk 3]
```diff
@@ -95,20 +100,31 @@ public static void closeQuietly(@Nullable ResultSet rs) {
...
+    return executeLargeInputs(input, function, i -> i);
+  }
+
+  /**
+   * Partition by 1000 elements a list of input and execute a function on each part.
+   *
+   * The goal is to prevent issue with ORACLE when there's more than 1000 elements in a 'in ('X', 'Y', ...)'
+   * and with MsSQL when there's more than 2000 parameters in a query
+   */
+  public static <OUTPUT, INPUT extends Comparable<INPUT>> List<OUTPUT> executeLargeInputs(Collection<INPUT> input, Function<List<INPUT>, List<OUTPUT>> function, IntFunction<Integer> partitionSizeManipulations) {
```
Comment (SonarTech) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: MINOR Split this 211 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java [hunk 4]
```diff
@@ -95,20 +100,31 @@ public static void closeQuietly(@Nullable ResultSet rs) {
...
+    return executeLargeInputs(input, function, size -> size == 0 ? Collections.emptyList() : new ArrayList<>(size), partitionSizeManipulations);
   }
 
-  public static <OUTPUT, INPUT extends Comparable<INPUT>> Set<OUTPUT> executeLargeInputsIntoSet(Collection<INPUT> input, Function<List<INPUT>, Set<OUTPUT>> function) {
-    return executeLargeInputs(input, function, size -> size == 0 ? Collections.emptySet() : new HashSet<>(size));
+  public static <OUTPUT, INPUT extends Comparable<INPUT>> Set<OUTPUT> executeLargeInputsIntoSet(Collection<INPUT> input, Function<List<INPUT>, Set<OUTPUT>> function,
```
Comment (SonarTech) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java [hunk 5]
```diff
@@ -124,7 +140,17 @@ public static void closeQuietly(@Nullable ResultSet rs) {
...
+    executeLargeUpdates(inputs, consumer, i -> i);
+  }
+
+  /**
+   * Partition by 1000 elements a list of input and execute a consumer on each part.
+   *
+   * The goal is to prevent issue with ORACLE when there's more than 1000 elements in a 'in ('X', 'Y', ...)'
+   * and with MsSQL when there's more than 2000 parameters in a query
+   */
+  public static <INPUT extends Comparable<INPUT>> void executeLargeUpdates(Collection<INPUT> inputs, Consumer<List<INPUT>> consumer, IntFunction<Integer> partitionSizeManipulations) {
```
Comment (SonarTech) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: MINOR Split this 183 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java [hunk 6]
```diff
@@ -138,10 +164,12 @@ public static void closeQuietly(@Nullable ResultSet rs) {
    *
    * @param sqlCaller a {@link Function} which calls the SQL update/delete and returns the number of updated/deleted rows.
    *
+   * @param partitionSizeManipulations
    * @return the total number of updated/deleted rows (computed as the sum of the values returned by {@code sqlCaller}).
    */
-  public static <INPUT extends Comparable<INPUT>> int executeLargeUpdates(Collection<INPUT> inputs, Function<List<INPUT>, Integer> sqlCaller) {
-    Iterable<List<INPUT>> partitions = toUniqueAndSortedPartitions(inputs);
+  public static <INPUT extends Comparable<INPUT>> int executeLargeUpdates(Collection<INPUT> inputs, Function<List<INPUT>, Integer> sqlCaller,
```
Comment (SonarTech) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java [hunk 7]
```diff
@@ -153,7 +181,15 @@ public static void closeQuietly(@Nullable ResultSet rs) {
...
    */
   public static <INPUT extends Comparable<INPUT>> Iterable<List<INPUT>> toUniqueAndSortedPartitions(Collection<INPUT> inputs) {
-    return Iterables.partition(toUniqueAndSortedList(inputs), PARTITION_SIZE_FOR_ORACLE);
+    return toUniqueAndSortedPartitions(inputs, i -> i);
+  }
+
+  /**
+   * Ensure values {@code inputs} are unique (which avoids useless arguments) and sorted before creating the partition.
+   */
+  public static <INPUT extends Comparable<INPUT>> Iterable<List<INPUT>> toUniqueAndSortedPartitions(Collection<INPUT> inputs, IntFunction<Integer> partitionSizeManipulations) {
```
Comment (SonarTech) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00119)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![MINOR][MINOR] 6 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java (hunk 5 above)
Comment (simonbrandhof) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: I suggest to complete javadoc with:
```
* @param input the whole list of elements to be partitioned
   * @param function the mapper method to be executed, for example {@code mapper(dbSession)::selectByUuids}
   * @param partitionSizeManipulations the function that computes the number of usages of a partition, for example
   *                                   {@code partitionSize -> partitionSize / 2} when the partition of elements
   *                                   in used twice in the SQL request.
```

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java [hunk 8]
```diff
@@ -138,10 +164,12 @@ public static void closeQuietly(@Nullable ResultSet rs) {
    *
    * @param sqlCaller a {@link Function} which calls the SQL update/delete and returns the number of updated/deleted rows.
    *
+   * @param partitionSizeManipulations
```
Comment (simonbrandhof) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: see previous comment

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java (hunk 3 above)
Comment (bartfastiel) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: Done.

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java (hunk 4 above)
Comment (bartfastiel) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: I want to keep `INPUT` and `OUTPUT` for better readability.

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java (hunk 5 above)
Comment (bartfastiel) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: Done.

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java (hunk 6 above)
Comment (bartfastiel) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: I want to keep `INPUT` and `OUTPUT` for better readability.

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java (hunk 7 above)
Comment (bartfastiel) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: I want to keep `INPUT` and `OUTPUT` for better readability.

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java (hunk 5 above)
Comment (bartfastiel) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: 👍  Done.

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java (hunk 8 above)
Comment (bartfastiel) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: Done.

File: server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java [hunk 9]
```diff
@@ -95,20 +100,32 @@ public static void closeQuietly(@Nullable ResultSet rs) {
...
+   *
+   * The goal is to prevent issue with ORACLE when there's more than 1000 elements in a 'in ('X', 'Y', ...)'
+   * and with MsSQL when there's more than 2000 parameters in a query
+   */
+  public static <OUTPUT, INPUT extends Comparable<INPUT>> List<OUTPUT> executeLargeInputs(Collection<INPUT> input, Function<List<INPUT>, List<OUTPUT>> function,
+    IntFunction<Integer> partitionSizeManipulations) {
+    return executeLargeInputs(input, function, size -> size == 0 ? Collections.emptyList() : new ArrayList<>(size), partitionSizeManipulations);
+  }
+
+  public static <OUTPUT, INPUT extends Comparable<INPUT>> Set<OUTPUT> executeLargeInputsIntoSet(Collection<INPUT> input, Function<List<INPUT>, Set<OUTPUT>> function,
```
Comment (SonarTech) [inline server/sonar-db-core/src/main/java/org/sonar/db/DatabaseUtils.java]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00119)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**